feat: add flag to set init policy to true

eduardacoppo opened the pull request:
https://github.com/rock-core/tools-syskit/pull/459 shouldn't have been merged. I was still working on a failed test.

eduardacoppo committed:
feat: use recommend_init flag to apply init policy

eduardacoppo committed:
chore: fix rubocop offense by moving code

fix Metrics/CyclomaticComplexity offense by moving the connection validation code to a separate function

eduardacoppo committed:
test: add tests for init policy

eduardacoppo committed:
fix/chore: policy for data readers and exported ports

and rewrite code in port model

eduardacoppo committed:
test: add tests and update existing ones

eduardacoppo committed:
chore: rewrite tests to call init_policy

eduardacoppo committed:
chore: remove unneccessary code in models/port and tests

eduardacoppo committed:
fix: do not override existing init flag

Use init_policy from the port as a default for the policy
Do not override the existing init flag in the policy
`policy.fetch(:init, port.model.init_policy)` to set the init value if not already provided in the policy

eduardacoppo committed:
fix/test: allow policy merge and add tests for this

eduardacoppo committed:
fix: do not merge if no init policy

eduardacoppo committed:
fix: fix merge test

doudou commented on lib/syskit/dynamic_port_binding.rb:
```diff
@@ -203,6 +203,9 @@ def initialize(
             # Method called by {Accessor} to create the accessor object from a
             # port
             def create_accessor(port)
+                if port.model.init_policy?
+                    policy = policy.merge(init: port.model.init_policy)
+                end
```
I don't remember how this was ... but

- it should fail with a nomethoderror (`policy` is nil in there), as in `policy.merge` refers to the local variable and not the method from the accessor.
- Unless I am mistaken, it should be model.init_policy?
- it overrides the explicit policy with the default, and it should be the other way around. That is, you should do `{ init: default_value }.merge(policy)` and not the other way.

You can use `@policy` to refer to the policy given to the constructor.

That's a lot of errors for 5 lines of code after a failed merge...

eduardacoppo replied:
> I don't remember how this was ...

These changes are new, previously I was always setting the policy.

Misunderstood you earlier, policy should keep being set. Got it.

doudou commented on lib/syskit/network_generation/dataflow_dynamics.rb:
```diff
@@ -620,10 +620,9 @@ def policy_for(
-                    policy.merge(init: source_port_m.init_policy)
-                end
-                policy
+                return policy unless source_port_m.init_policy?
```
Once more. Always set the policy. It makes no sense to not set it to false in this case.

doudou commented on lib/syskit/dynamic_port_binding.rb:
```diff
@@ -203,6 +203,7 @@ def initialize(
             # Method called by {Accessor} to create the accessor object from a
             # port
             def create_accessor(port)
+                @policy.merge(init: @policy.fetch(:init, port.model.init_policy?))
```
- `merge` is returning a new hash. This line does nothing.
- the best way to control priority with `merge` is to decide what hash gets merged into which, essentially: `low_priority.merge(high_priority)`

doudou commented on test/test_dynamic_port_binding.rb:
```diff
@@ -382,6 +382,45 @@ module Syskit
+                @task.out_port.model.init_policy(true)
+                flexmock(@task.out_port)
+                    .should_receive(:reader)
+                    .and_return({ init: true })
```
Should be `with` and not `and_return`. Which is why the tests pass even though the implementation is wrong.

eduardacoppo committed:
fix: assign merge to policy and fix priority

- Correctly merges policy, making sure :init is only set if it wasn't already defined.
  This prevents :init from being overwritten when explicitly provided.
- Adds a test to confirm that an existing :init value remains unchanged.

eduardacoppo committed:
chore: fix rubocop offense

eduardacoppo committed:
fix/test: fix policy handling in tests

Fixes incorrect .and_return(...) usage in tests, replacing it with .with(...) to properly validate method arguments.

doudou commented on test/network_generation/test_dataflow_dynamics.rb:
```diff
@@ -380,15 +420,53 @@ module NetworkGeneration
+                    flexmock(@dynamics)
+                        .should_receive(:compute_reliable_connection_policy)
+                        .with(@source_t.out_port, @sink_t.in_port, fallback_policy)
+                        .once.and_return({ init: true })
```
*You* are returning `init: true` from `compute_reliable_connection_policy`. Not sure what it is meant to test.

jhonasiv commented on test/network_generation/test_dataflow_dynamics.rb:
```diff
@@ -380,15 +420,53 @@ module NetworkGeneration
+                    flexmock(@dynamics)
+                        .should_receive(:compute_reliable_connection_policy)
+                        .with(@source_t.out_port, @sink_t.in_port, fallback_policy)
+                        .once.and_return({ init: true })
```
```suggestion
                        .once
```

doudou replied:
She still needs to return something.

jhonasiv replied:
Also talked in-person, we decided that in here an empty hash works

jhonasiv commented on test/network_generation/test_dataflow_dynamics.rb:
```diff
@@ -380,15 +420,53 @@ module NetworkGeneration
+
+                    flexmock(@source_t.out_port.model)
+                        .should_receive(:init_policy?).explicitly
+                        .and_return(true)
```
```suggestion

```

doudou replied:
??? In that case, it makes sense, no ?

jhonasiv replied:
Yeah, I talked with her in person and it did make sense, forgot to update the review in here

jhonasiv commented on test/network_generation/test_dataflow_dynamics.rb:
```diff
@@ -380,15 +420,53 @@ module NetworkGeneration
+                    @source_t.out_port.model.init_policy(true)
+                    policy = @dynamics.policy_for(@source_t, "out", "in", @sink_t, nil)
+
+                    assert_equal true, policy[:init]
```
```suggestion
                    assert policy[:init]
```

doudou commented on test/network_generation/test_dataflow_dynamics.rb:
```diff
@@ -380,15 +420,51 @@ module NetworkGeneration
+
+                    flexmock(@source_t.out_port.model)
+                        .should_receive(:init_policy?).explicitly
+                        .and_return({})
```
`init_policy?` returns `true` or `false`. The `true` here was correct.

What you're doing with `and_return({})` is put a "true-ish" value.
